### PR TITLE
CI experience

### PR DESCRIPTION
**Pull request by Bindsi**
- use of deislab´s official shim image with latest MS redis trigger to create the k3d cluster
- removed deislab subtree (not needed anymore due to the latest integration of Suneet´s redis trigger dev)
- added workflow to execute the integration test
- added workflow to push the apps images into the GHCR
- added docs to describe the CI process and to explain prerequisites

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,21 +1,19 @@
...
+K3DSHIMIMAGENAME := ghcr.io/deislabs/containerd-wasm-shims/examples/k3d
 DOCKERDIR := ./wasm-shims/deployments/k3d
-TMPSPINDIR := ./wasm-shims/deployments/k3d/.tmp
 APPSDIR := ./apps
+GITHUBORG := azure-samples
+GITHUBREPO := wasm-orchestration-with-spin
 
 all: build_k3d_node_image create_k3d_cluster install_redis deploy_app run_integrationtest
 
 build_k3d_node_image:
```
[the remark this answers is not in the document]
Comment (suneetnangia) [reply Makefile:10]: Do we need to pull the node image beforehand? or can `k3d cluster create` cmd download it directly?

File: docs/ci.md
Comment (suneetnangia) [inline docs/ci.md:1]: nit: run spell checker (it should be installed in devcontainer already).

File: docs/ci.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Continuous Integration
```
Comment (suneetnangia) [inline docs/ci.md:1]: nit: add this `Continuous Integration (CI)` as title so CI acronym is explained already.

File: docs/ci.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+This documents describes the CI process for the solution - in partucular for the Spin apps.
+
+## Integration Tests
+
+The workflow [PR](../.github/workflows/PR.yaml) is used as a validation pipeline for pull requests created against the main branch.
+
+The job contains the generation and setup of the devcontainer including the docker-in-docker installation. That is required to build the wasm apps because it can be configured to use the containerd runtime for pulling and storing images as explaining in the [Docker-In-Docker](../docs/ci.md#docker-in-docker) section.
```
Comment (suneetnangia) [inline docs/ci.md:9]: `use the containerd runtime` -> `use the wasm based containerd runtime`

File: docs/ci.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+The pipeline also configures the k3d cluster and deploys the apps into the cluster. The apps are validated by running the integration tests as part of the makefile experience.
+
+## Release Pipeline
+
+The workflow [Release](../.github/workflows/Release.yaml) is triggered when check-in into the main branch is commited (ideally as part of the PR merge) and is used to push the apps images into the GitHub Container Registry (GHCR).
```
Comment (suneetnangia) [inline docs/ci.md:14]: `when check-in into`  can be rephrased?

File: docs/ci.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+## Configuration
+
+### GITHUB_TOKEN
+
+The GITHUB_TOKEN secret is required to be configured with "packages write" permission to be able to push the images into the GHCR from the Docker-In-Docker container.
+
+```yaml
+permissions:
+    packages: write
```
Comment (suneetnangia) [inline docs/ci.md:28]: nit: what does this do? I wonder if adding helpful comments in yaml itself would be better option, so readers do not have to do back and forth?

File: docs/ci.md
Comment (suneetnangia) [inline docs/ci.md:1]: I think in general, having a lot these comments in actual files e.g. yamls would be easier for readers to understand.

**Review by suneetnangia — COMMENTED** [review]
Significant CI work in this PR, thank you. Added some comments for your review.

File: .github/workflows/Release.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        docker run --name devcontainer -d --rm --privileged --mount type=bind,source=${{ github.workspace }},target=/workspace devcontainer
+    - name: install tools
+      run: |
+        docker exec devcontainer sh ./workspace/.devcontainer/bootstrap.sh
+    - name: log in to the Container registry
+      run: |
+        docker exec devcontainer sh -c "docker login ghcr.io -u ${{ github.actor }} -p ${{ secrets.GITHUB_TOKEN }}"
+    - name: push workload images
+      run: |
+        docker exec devcontainer sh -c "export PATH="/root/.cargo/bin:$PATH" && cd ./workspace && make build_push_app_images GITHUBORG=${{ github.repository_owner }} GITHUBREPO=${{ github.repository }}"
```
[the remark this answers is not in the document]
Comment (Bindsi) [reply .github/workflows/Release.yaml:36]: I´ll comment the yaml regarding it

File: .github/workflows/Release.yaml [hunk 7]
```diff
@@ -0,0 +1,36 @@
+name: Main - Upload Apps Images/Packages
+
+on: 
+  push:
+    branches: 
+      - main
+    paths:
+      # Only run this when apps are updated.
+      - 'apps/**'
```
[the remark this answers is not in the document]
Comment (Bindsi) [reply .github/workflows/Release.yaml:9]: agree

File: Makefile [hunk 8]
```diff
@@ -1,21 +1,19 @@
 K3DCLUSTERNAME := wasm-cluster
+K3DSHIMIMAGENAME := ghcr.io/deislabs/containerd-wasm-shims/examples/k3d
```
[the remark this answers is not in the document]
Comment (Bindsi) [reply Makefile:2]: makes totally sense

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (Bindsi) [reply Makefile:10]: those 2 vars GITHUBORG, GITHUBREPO are obsolete...forgot to remove them...we use the GH default vars now!!!

k3d cluster create can download...just tried...so we don´t need to pull anymore and can remove the step, but keeping the var on top for better handling of the image

File: docs/ci.md
Comment (Bindsi) [reply docs/ci.md:1]: added

File: docs/ci.md (hunk 2 above)
Comment (Bindsi) [reply docs/ci.md:1]: done

File: Makefile [hunk 9]
```diff
@@ -1,21 +1,19 @@
...
 DOCKERDIR := ./wasm-shims/deployments/k3d
```
[the remark this answers is not in the document]
Comment (Bindsi) [reply Makefile:3]: keep the var for better version and image handling

File: .github/workflows/PR.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    - name: run the devcontainer image
+      run: |
+        cd ..
+        docker run --name devcontainer -d --rm --privileged --mount type=bind,source=${{ github.workspace}},target=/workspace devcontainer
+    - name: install tools
+      run: |
+        docker exec devcontainer sh ./workspace/.devcontainer/bootstrap.sh
+    - name: install cluster and workloads
+      run: |
+        docker exec devcontainer sh -c "export PATH="/root/.cargo/bin:$PATH" && cd ./workspace && make"
```
[the remark this answers is not in the document]
Comment (Bindsi) [reply .github/workflows/PR.yaml:26]: added fmt target in makefile and call it before installing the cluster

File: .github/workflows/Release.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  Release-Apps-Images:
+    permissions:
+      packages: write
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+      name: download code
+      with:
+        fetch-depth: 0
+    - name: build the devcontainer image
```
[the remark this answers is not in the document]
Comment (suneetnangia) [reply .github/workflows/Release.yaml]: Move to .github/actions/build-devcontainer/action.yaml.

[the remark this answers is not in the document]
Comment (Bindsi) [reply .github/workflows/Release.yaml]: thanks for resolving

File: docs/ci.md
Comment (Bindsi) [reply docs/ci.md:1]: done

File: docs/ci.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
+# Continuous Integration (CI)
+
+This documents describes the CI process for the solution - in partucular for the Spin apps.
```
[the remark this answers is not in the document]
Comment (Bindsi) [reply docs/ci.md:3]: done

File: docs/ci.md (hunk 4 above)
Comment (Bindsi) [reply docs/ci.md:14]: done: when code is committed or merged into main branch

File: docs/ci.md (hunk 3 above)
Comment (Bindsi) [reply docs/ci.md]: done

**Review by suneetnangia — APPROVED** [review]
Brill, thanks.

**Pull request merged**